File: backend/services/expense_classifier.py
"""
Finansal sınıflandırma servisi.

Muhasebe prensiplerine uygun şekilde işlemleri 3 sınıfa ayırır:

1. VEHICLE_COST  → Araç maliyetine eklenir. Öz sermayeyi azaltmaz.
                    Satışta brüt kâr hesabında düşülür.
                    Örnek: Araç Alımı, Sahibine Ödeme, Ekspertiz, Noter, Çekici,
                    Bakım, Detaylı Yıkama, Plaka, Boyacı vb.

2. OPERATING     → İşletme gideridir. Anında öz sermayeyi azaltır.
                    Örnek: Kira, Maaş, SGK, Reklam, Vergi, Elektrik, İnternet,
                    Muhasebe, Ofis Gideri, Sigorta, Temizlik.

3. NEUTRAL       → Sermayeyi etkilemeyen, denetim amaçlı kategoriler.
                    Örnek: Kapora İadesi (alıcıya iade), Çalışan Payı,
                    Kapora Alındı (gelir tarafında, satış olmadığı için kâra eklenmez).

Bu sınıflandırma hem `transactions.expense_type` migrasyonu için, hem de
runtime `/api/finance/summary` hesaplamaları için tek doğruluk kaynağıdır.
"""
from typing import Literal

ExpenseClass = Literal["vehicle_cost", "operating", "neutral"]

# Araç maliyetine eklenen kategoriler (varlık dönüşümü)
VEHICLE_COST_CATEGORIES = {
    "Araç Alımı",
    "Araç Sahibine Ödeme",
    "Ekspertiz",
    "Noter",
    "Çekici",
    "Taşıma/Çekici",
    "Nakliye",
    "Bakım",
    "Bakım/Onarım",
    "Detaylı Yıkama",
    "Hazırlık",
    "Plaka",
    "Boyacı",
    "Kaporta",
    "Lastik",
    "Yedek Parça",
}

# İşletme gideri kategorileri (anlık sermaye azaltıcı)
OPERATING_CATEGORIES = {
    "Personel Maaşı",
    "Maaş",
    "SGK",
    "Kira",
    "Ofis Gideri",
    "Elektrik/Su/Doğalgaz",
    "İnternet/Telefon",
    "Sigorta",
    "Vergi",
    "Reklam/Pazarlama",
    "Reklam",
    "Muhasebe",
    "Temizlik",
    "Kırtasiye",
    "Yemek/İkram",
    "Ulaşım",
    "Diğer Gider",
    "Diğer",
}

# Sermayeyi etkilemeyen / iade-niteliğinde kategoriler
NEUTRAL_CATEGORIES = {
    "Kapora İadesi",
    "Çalışan Payı",
}


def classify_expense(category: str | None, car_id: str | None = None) -> ExpenseClass:
    """
    Bir gider tx'inin sınıfını döndür.

    Heuristik:
    - Açıkça VEHICLE_COST kategorisindeyse → vehicle_cost
    - Açıkça OPERATING kategorisindeyse  → operating
    - NEUTRAL ise → neutral
    - Bilinmeyen ama car_id'ye bağlıysa → vehicle_cost (araca özel masraf varsayımı)
    - Diğer her durumda → operating (güvenli default, çünkü işletme giderlerini
      eksik göstermek kullanıcıyı yanıltmaktan beter; fazla göstermek sermayeyi
      sadece olduğundan kötümser gösterir)
    """
    cat = (category or "").strip()
    if cat in VEHICLE_COST_CATEGORIES:
        return "vehicle_cost"
    if cat in OPERATING_CATEGORIES:
        return "operating"
    if cat in NEUTRAL_CATEGORIES:
        return "neutral"
    # Bilinmeyen kategori — car_id'ye bağlıysa araç maliyeti say
    if car_id:
        return "vehicle_cost"
    return "operating"


__all__ = [
    "VEHICLE_COST_CATEGORIES",
    "OPERATING_CATEGORIES",
    "NEUTRAL_CATEGORIES",
    "classify_expense",
    "ExpenseClass",
]

File: frontend/src/utils/expenseClassifier.js
/**
 * Frontend kategori sınıflandırma — backend services/expense_classifier.py ile birebir uyumlu.
 *
 * Single source of truth: backend tarafında karar verilir. Burası sadece
 * raporların ve modalların runtime sınıflandırma yapması için duplicate edildi.
 * Yeni kategori eklenirse iki tarafta da güncellenmeli.
 */

export const VEHICLE_COST_CATEGORIES = new Set([
  'Araç Alımı',
  'Araç Sahibine Ödeme',
  'Ekspertiz',
  'Noter',
  'Çekici',
  'Taşıma/Çekici',
  'Nakliye',
  'Bakım',
  'Bakım/Onarım',
  'Detaylı Yıkama',
  'Hazırlık',
  'Plaka',
  'Boyacı',
  'Kaporta',
  'Lastik',
  'Yedek Parça',
]);

export const OPERATING_CATEGORIES = new Set([
  'Personel Maaşı',
  'Maaş',
  'SGK',
  'Kira',
  'Ofis Gideri',
  'Elektrik/Su/Doğalgaz',
  'İnternet/Telefon',
  'Sigorta',
  'Vergi',
  'Reklam/Pazarlama',
  'Reklam',
  'Muhasebe',
  'Temizlik',
  'Kırtasiye',
  'Yemek/İkram',
  'Ulaşım',
  'Diğer Gider',
  'Diğer',
]);

export const NEUTRAL_CATEGORIES = new Set([
  'Kapora İadesi',
  'Çalışan Payı',
]);

/**
 * Bir gider tx'inin sınıfını döndür.
 * @returns {'vehicle_cost' | 'operating' | 'neutral'}
 */
export function classifyExpense(category, carId = null) {
  const cat = (category || '').trim();
  if (VEHICLE_COST_CATEGORIES.has(cat)) return 'vehicle_cost';
  if (OPERATING_CATEGORIES.has(cat)) return 'operating';
  if (NEUTRAL_CATEGORIES.has(cat)) return 'neutral';
  if (carId) return 'vehicle_cost';
  return 'operating';
}

export function isOperatingExpense(tx) {
  if (!tx || tx.type !== 'expense') return false;
  return classifyExpense(tx.category, tx.car_id) === 'operating';
}

export function isVehicleCost(tx) {
  if (!tx || tx.type !== 'expense') return false;
  return classifyExpense(tx.category, tx.car_id) === 'vehicle_cost';
}
